lib/src/piece/list.dart

```diff
@@ -11,8 +11,17 @@ import 'piece.dart';
 /// class handles adding and removing the trailing comma depending on whether
 /// the list is split or not.
 ///
-/// Usually constructed using a [DelimitedListBuilder].
+/// Usually constructed using [createDelimited()] or a [DelimitedListBuilder].
 class ListPiece extends Piece {
+  /// State used for a split list containing type arguments or type parameters.
+  ///
+  /// Has a higher cost than [State.split] since splitting type arguments and
+  /// type parameters tends to look worse than splitting at other places.
+  // TODO(rnystrom): Having to use a different state for this is a little
+  // cumbersome. Maybe it would be better to most costs out of [State] and
+  // instead have the [Solver] ask each [Piece] for the cost of its state.
+  static const _splitTypes = State(1, cost: 2);
```

Comment:
Would it make more sense for this to be in `State` instead of `ListPiece`? To keep it together.
If not, maybe name it `_splitTypesState`?

It seems interesting that cost has been just a number to be assigned to a `State` haha. Maybe there's some other reasoning I don't understand. Would you have to update this if you're adding more states with different costs in between other types of states.

Reply:
> Would it make more sense for this to be in `State` instead of `ListPiece`? To keep it together.
> If not, maybe name it `_splitTypesState`?

I'm honestly not sure. I've gone back and forth on this question in my mind. If you look in the old IR, the costs used by all the different rules are [in one file](https://github.com/dart-lang/dart_style/blob/main/lib/src/constants.dart). That can help you see how the various constructs have costs relative to each other. On the other hand, it means the costs are not as encapsulated in the rules that actually use them.

I'm not sure what the right trade-off is yet. In practice, it probably doesn't matter too much one way or the other. One nice thing about keeping the piece-specific states in their respective pieces is that I think it makes it easier to document them because the reader already understands the kind of syntax the state is being used to format.

> It seems interesting that cost has been just a number to be assigned to a `State` haha.

I'm not sold on that design either. In the old IR, each Rule (which is roughly analogous to a Piece) [knows its cost](https://github.com/dart-lang/dart_style/blob/main/lib/src/rule/rule.dart#L47). I'm considering moving cost out of State and instead having a method on Piece where you can ask it "what should the cost of this state be":

```dart
int stateCost(State state);
```

That would give pieces more fine-grained control to choose different costs for states that are used by multiple pieces. Also, it would eliminate the need to create different states that represent conceptually the same thing but just with a different cost.

None of this is carved in stone yet. :)

> Would you have to update this if you're adding more states with different costs in between other types of states.

Costs don't have to be unique, so that's fine. We rely on distinguishable states having distinct `_value`s for that.